### ISSUE TITLE
Feature request: inner xsep west/east and inner ysep above/below

### ISSUE THREAD

**Issue by pgf-tikz-bot** [pgf-tikz/pgf#631]
*Migrated from SourceForge*
Author: carlatex
Timestamp: 2017-07-01 18:35:29.759000

At present, you can set the inner sep globally, for both vertical borders or for the both horizontal borders, for example:

```
\documentclass{article}
\usepackage{tikz}
\usetikzlibrary{positioning}

\begin{document}
	\begin{tikzpicture}
		\node[draw] (A) {\colorbox{red}{Normal node}};
		\node[draw, below=of A, inner sep=0pt] (B) {\colorbox{red}{Node with inner sep=0pt}};
		\node[draw, below=of B, inner xsep=0pt] (C) {\colorbox{red}{Node with inner xsep=0pt}};
		\node[draw, below=of C, inner ysep=0pt] {\colorbox{red}{Node with inner ysep=0pt}};
	\end{tikzpicture}
\end{document}
```

Is it possible to have an `inner xsep` only for west or east side (i.e. `inner xsep west` and `inner xsep east`), and an `inner ysep` only for the upper or lower border (i.e. `inner ysep above` and `inner ysep below`)?

See also this post on TeX.SE: https://tex.stackexchange.com/q/27893/101651

**Comment by hmenke** [comment]
@CarLaTeX I'm not sure that makes sense for all shapes, so I guess you only want it for `rectangle`.

**Comment by CarLaTeX** [comment]
Yes, you're right. It makes sense only for rectangle.
Thank you for the great work you're doing!
Ciao!
Carla

Il Ven 5 Apr 2019, 06:11 Henri Menke <notifications@github.com> ha scritto:

**Comment by hmenke** [comment]
**tl;dr** I had a quick look.  There are two options.  Neither of them is good.  Unless you have a better one, I unfortunately have to reject this feature request.

### Option 1

The most sensible way to implement this would be to make `inner xsep` a proxy for `inner xsep west` and `inner xsep east` like so
```latex
\pgfset{
  inner xsep east/.initial  =.3333em,
  inner xsep west/.initial  =.3333em,
  inner xsep/.style         ={/pgf/inner xsep east=#1,/pgf/inner xsep west=#1},
}
```
Currently a lot of node drawing code uses constructions like this
```latex
\pgfmathsetlength\pgf@x{\pgfkeysvalueof{/pgf/inner xsep}}%
```
which would naturally break, because `inner xsep` no longer holds a value.

Now I could go ahead an rewrite all of the node drawing code to use the correct `inner xsep west` and `inner xsep east` respectively.  This will be a huge lot of work which I don't know whether I am willing to do.

Another problem is that PGF actually has clients, i.e. other packages that depend on it.  Quickly searching the TEXMF tree yields:
```
/usr/local/texlive/2019/texmf-dist/tex/latex $ ag '\\pgfkeysvalueof{/pgf/inner xsep}'
tikz-sfc/tikzlibrarycircuits.plc.sfc.code.tex
61:      {\pgfpointadd{\southwest}{\pgfpoint{6*\pgfkeysvalueof{/pgf/inner xsep}}{6*\pgfkeysvalueof{/pgf/inner ysep}}}}
62:      {\pgfpointadd{\northeast}{\pgfpointscale{-1}{\pgfpoint{6*\pgfkeysvalueof{/pgf/inner xsep}}{6*\pgfkeysvalueof{/pgf/inner ysep}}}}}

tikz-ladder/tikzlibrarycircuits.plc.ladder.code.tex
167:  \pgfmathaddtolength\pgf@x{\pgfkeysvalueof{/pgf/inner xsep}}
340:			\pgfmathaddtolength\pgf@xc{\pgfkeysvalueof{/pgf/inner xsep}}
375:			\pgfmathaddtolength\pgf@xc{-\pgfkeysvalueof{/pgf/inner xsep}}

spectralsequences/sseqdrawing.code.tex
1249:    \pgfmathsetlength\pgf@xb{\pgfkeysvalueof{/pgf/inner xsep}}%
1327:    \pgfmathsetlength\pgf@xb{\pgfkeysvalueof{/pgf/inner xsep}}%

pgf/compatibility/pgfcomp-version-1-18.sty
21:\def\pgfshapeinnerxsep{\pgfkeysvalueof{/pgf/inner xsep}}

forest/forest.sty
4034:        ($(.anchor)+(\forestoption{min x}+\pgfkeysvalueof{/pgf/outer xsep}/2+\pgfkeysvalueof{/pgf/inner xsep},\forestoption{min y}+\pgfkeysvalueof{/pgf/outer ysep}/2+\pgfkeysvalueof{/pgf/inner ysep})$)
4036:        ($(.anchor)+(\forestoption{max x}-\pgfkeysvalueof{/pgf/outer xsep}/2-\pgfkeysvalueof{/pgf/inner xsep},\forestoption{max y}-\pgfkeysvalueof{/pgf/outer ysep}/2-\pgfkeysvalueof{/pgf/inner ysep})$)
8743:    \pgfkeysvalueof{/pgf/inner xsep},\pgfkeysvalueof{/pgf/outer xsep}%
8748:    s sep'/.expanded={\the\dimexpr \pgfkeysvalueof{/pgf/inner xsep}*2}
```
So already four packages would break when I change only `inner xsep` (this doesn't yet account for `inner ysep`, `outer xsep` and `outer ysep`).

That effectively means that this option is dead.

### Option 2

The other option would be to just make `inner xsep east` and `inner xsep west` completely independent options.
```latex
\pgfset{
  inner xsep east/.initial  =.3333em,
  inner xsep west/.initial  =.3333em,
  inner xsep/.style         =.3333em,
}
```
Now we have a problem of synchronization. Every node drawing code that is supposed to respect `inner xsep east` and `inner xsep west` now requires branching code to detect whether they have been set independently from `inner xsep`.  There are also corner cases where the behaviour is undefined, like this
```latex
\node[draw, inner sep=0pt, inner xsep east=10pt, inner xsep west=5pt] {Hallo!};
```
One could argue that the last one to be set should take precedence, i.e. `inner sep=0pt` should be ignored but that particular state of the parser is not accessible from within the node drawing code.  

On top of undefined behaviour you will also get unexpected behaviour.  Let's say `inner xsep east` and `inner xsep west` haven't been implemented for the `circle` shape, then
```latex
\node[circle, draw, inner xsep east=10pt] {Hallo!};
```
will just silently ignore `inner xsep east`.

Since this option is highly inconsistent and will just give me more bug reports, I reject it right away.

**Comment by CarLaTeX** [comment]
Ok, I thought it was simpler.
Thank you, anyway.
Ciao!
Carla

Il Ven 5 Apr 2019, 08:12 Henri Menke <notifications@github.com> ha scritto:

**Comment by hmenke** [comment]
As you can see, option 1 is conceptually pretty simple, but I don't want to break every package that depends on PGF.  Sorry :(

**Comment by CarLaTeX** [comment]
Yes, of course! Ciao!

Il Ven 5 Apr 2019, 08:23 Henri Menke <notifications@github.com> ha scritto:

**Comment by aubertc** [comment]
As [hinted here](https://tex.stackexchange.com/a/517227/34551), you can play with `yshift` and `xshift` to obtain asymmetrical inner seps.
